File: taskActivation/stimuli_ptb/checkerboard_3cond_task.m
function checkerboard_3cond_task(varargin)
%CHECKERBOARD_3COND_TASK Psychtoolbox presentation of a 3-position flickering-
%   checkerboard visual localizer, flickering in one of three screen
%   positions per block: CENTER, LEFT, or RIGHT -- OFF (blank) -> ON
%   (pattern A) -> OFF -> ON (pattern B, the black<->white inverse of A)
%   -> repeat, at 'FlickerHz'. CENTER is a SQUARE (its height set equal to
%   its own width, not the window height) centered on screen, so it
%   stimulates only the fovea; LEFT and RIGHT are full window-height BARS,
%   anchored flush against the window's left/right edge respectively (not
%   floating with a gap), so they reach as far into the visual periphery
%   as the window allows. A small + fixation cross stays visible at screen
%   center THROUGHOUT every block (including LEFT/RIGHT, and every OFF
%   phase), so the subject can hold central gaze while the checkerboard
%   stimulates each position -- standard practice for a peripheral visual
%   localizer, so the resulting activation reflects retinotopic stimulus
%   location rather than eye movements.
%
%   Shows a brief task-instructions screen (dismiss with 1/2, or Escape to
%   abort), waits for the scanner trigger, then presents
%   ../study_design/Checkerboard3Cond_events.tsv (rest, center, rest, left,
%   rest, right, x4 + a trailing rest block; 12s blocks, 25 blocks, 300s
%   total). Matches conf/checkerboard_3cond.toml's 3-way one-vs-rest GLM
%   contrast (glmCondA=center / glmCondB=left / glmCondC=right, each
%   contrasted against the mean of the other two) for live analysis.
%
%   Name-value options (all optional):
%     'TriggerKey'    cellstr of trigger keys (default {'5','5%','t'} -- '5%'
%                     covers sites where the scanner's sync pulse arrives as
%                     the shifted-symbol name PTB gives that key; run
%                     test_ptb_install.m and press your actual trigger/box
%                     once to see exactly what name it reports if unsure)
%     'EventsFile'    path to events.tsv (default: Checkerboard3Cond_events.tsv)
%     'Duration'      total run length in seconds (default: end of the last
%                     event); pass nVols*TR to also show trailing rest
%     'FlickerHz'     how many times per second the checkerboard's state
%                     changes during a CENTER/LEFT/RIGHT block (default 8):
%                     OFF (blank) -> ON (pattern A) -> OFF -> ON (pattern
%                     B, the black<->white inverse of A) -> repeat, each
%                     state lasting 1/FlickerHz -- same semantics as
%                     checkerboard_1cond_task.m
%     'Windowed'      true for a windowed test window (default false)
%     'ScreenWidth'   \
%     'ScreenHeight'   } pixel resolution to open at, e.g. 1920/1080 for a
%                     known scanner-room projector (default: [], auto-detect
%                     the display's own native resolution -- see
%                     ptb_open_window.m). Ignored if 'Windowed' is true.
%     'SkipSyncTests' true to disable PTB's flip-timing sync tests, for
%                     testing on a non-research display (default false)
%     'LogPath'       timing log path (default:
%                     stimuli_ptb/logs/checkerboard_3cond_task_<timestamp>.tsv)
%
%   Example:
%     checkerboard_3cond_task('Windowed', true, 'TriggerKey', {'space'})   % test
%     checkerboard_3cond_task('ScreenWidth', 1920, 'ScreenHeight', 1080)   % real run

    here = fileparts(mfilename('fullpath'));

    p = inputParser;
    addParameter(p, 'TriggerKey', {'5', '5%', 't'});
    addParameter(p, 'EventsFile', fullfile(here, '..', 'study_design', 'Checkerboard3Cond_events.tsv'));
    addParameter(p, 'Duration', []);
    addParameter(p, 'FlickerHz', 8);
    addParameter(p, 'Windowed', false);
    addParameter(p, 'ScreenWidth', []);
    addParameter(p, 'ScreenHeight', []);
    addParameter(p, 'SkipSyncTests', false);
    addParameter(p, 'LogPath', '');
    parse(p, varargin{:});
    opt = p.Results;

    logPath = opt.LogPath;
    if isempty(logPath)
        logPath = fullfile(here, 'logs', ...
            sprintf('checkerboard_3cond_task_%s.tsv', datestr(now, 'yyyymmdd_HHMMSS')));
    end

    events = ptb_read_events_tsv(opt.EventsFile);
    fprintf('[checkerboard_3cond_task] loaded %d events from %s\n', height(events), opt.EventsFile);

    screenSize = [];
    if ~isempty(opt.ScreenWidth) && ~isempty(opt.ScreenHeight)
        screenSize = [opt.ScreenWidth, opt.ScreenHeight];
    end
    win = ptb_open_window(opt.Windowed, opt.SkipSyncTests, [], screenSize);
    cleanupWin = onCleanup(@() sca);   % guarantees the display is released on ANY exit --
                                       % normal completion, Escape-abort, or an uncaught error

    % ---- build the two phase-inverted checkerboard textures for CENTER
    %      and for LEFT/RIGHT (which share the same width, so share one
    %      texture pair) up front. Each texture's own pixel grid is sized
    %      to match that stim's real width/height in squareSizePx cells, so
    %      the checker cells are genuinely SQUARE regardless of the stim's
    %      own aspect ratio -- rather than stretching one shared square
    %      texture into a non-square destRect, which would elongate the
    %      cells. ----
    squareSizePx = 40;
    [screenW, screenH] = Screen('WindowSize', win);
    shortSide = min(screenW, screenH);

    centerWidthFraction = 0.288; % fraction of the screen's shorter
                                 % dimension -- CENTER's own WIDTH, also
                                 % used as its HEIGHT (a square -- see
                                 % destRects below); LEFT/RIGHT bars span
                                 % the full screen HEIGHT instead
    sideWidthFraction = 0.32;    % still smaller than centerWidthFraction
                                 % so LEFT/RIGHT stay visually distinct
                                 % from CENTER

    barWidthCenter = shortSide * centerWidthFraction;   % CENTER's own width AND height
    barWidthSide = shortSide * sideWidthFraction;

    nColsCenter = max(2, round(barWidthCenter / squareSizePx));
    nRowsCenter = nColsCenter;   % CENTER is a square, so rows == cols
    nColsSide = max(2, round(barWidthSide / squareSizePx));
    nRowsSide = max(2, round(screenH / squareSizePx));

    [Xc, Yc] = meshgrid(0:nColsCenter - 1, 0:nRowsCenter - 1);
    checkerCenter = mod(Xc + Yc, 2);
    texCenterA = Screen('MakeTexture', win, uint8(checkerCenter * 255));
    texCenterB = Screen('MakeTexture', win, uint8((1 - checkerCenter) * 255));

    [Xs, Ys] = meshgrid(0:nColsSide - 1, 0:nRowsSide - 1);
    checkerSide = mod(Xs + Ys, 2);
    texSideA = Screen('MakeTexture', win, uint8(checkerSide * 255));
    texSideB = Screen('MakeTexture', win, uint8((1 - checkerSide) * 255));

    % CENTER is a SQUARE centered on screen (stimulates only the fovea);
    % LEFT/RIGHT span the FULL window height and are anchored flush
    % against the window's left/right edge (no margin/gap), so each
    % reaches as far into the periphery as the window allows.
    destRects = struct( ...
        'center', CenterRectOnPointd([0 0 barWidthCenter barWidthCenter], ...
            screenW / 2, screenH / 2), ...
        'left', [0, 0, barWidthSide, screenH], ...
        'right', [screenW - barWidthSide, 0, screenW, screenH]);

    texA = struct('center', texCenterA, 'left', texSideA, 'right', texSideA);
    texB = struct('center', texCenterB, 'left', texSideB, 'right', texSideB);
    flickerHz = opt.FlickerHz;

    function stimFor(win, trialType, tInEvent, ~)
        if isfield(destRects, trialType)
            % True flicker: fully OFF (blank -- nothing drawn, so the
            % window's black background shows through) between each ON
            % flash, and the ON flash itself alternates pattern A/B
            % (black<->white inverse) so a given screen location genuinely
            % reverses polarity from one flash to the next -- same 4-phase
            % cycle as checkerboard_1cond_task.m.
            phase = mod(floor(tInEvent * flickerHz), 4);
            if phase == 1
                Screen('DrawTexture', win, texA.(trialType), [], destRects.(trialType));
            elseif phase == 3
                Screen('DrawTexture', win, texB.(trialType), [], destRects.(trialType));
            end
            % phase == 0 or 2: blank -- draw nothing.
        end
        % Fixation stays visible for EVERY frame of EVERY condition
        % (including rest) -- see the header comment on why: keeps gaze
        % central while LEFT/RIGHT stimulate the periphery. Green (not
        % white) so it stays clearly visible against the black-and-white
        % checkerboard squares behind it.
        DrawFormattedText(win, '+', 'center', 'center', [0 1 0]);
    end

    instructions = ['CHECKERBOARD LOCALIZER TASK\n\n' ...
        'You will see a flickering black-and-white checkerboard pattern appear in ' ...
        'different places on the screen -- sometimes in the middle, sometimes off ' ...
        'to the left, sometimes off to the right.\n\n' ...
        'Goal: keep looking at the + in the center of the screen THE WHOLE TIME, ' ...
        'even when the checkerboard appears off to one side -- just notice it out ' ...
        'of the corner of your eye rather than looking directly at it. No response ' ...
        'or button press is needed during the task.\n\n' ...
        'If you have any questions, ask the experimenter now. When you are ' ...
        'comfortable, press any button to continue.'];
    if ptb_show_instructions(win, instructions)
        return
    end

    t0 = ptb_wait_for_trigger(win, opt.TriggerKey, 'Waiting for scanner trigger...');
    fprintf('[checkerboard_3cond_task] triggered at t0=%.3fs -- starting task\n', t0);
    ptb_run_block_loop(win, events, @stimFor, t0, opt.Duration, logPath);

    DrawFormattedText(win, 'Task complete -- thank you!', 'center', 'center', [1 1 1]);
    Screen('Flip', win);
    WaitSecs(2.0);
end

File: taskActivation/stimuli/checkerboard_3cond_task.py
#!/usr/bin/env python3
"""-----------------------------------------------------------------------------
checkerboard_3cond_task.py -- simple PsychoPy presentation of this project's
own 3-position flickering-checkerboard visual localizer
(study_design/Checkerboard3Cond_events.tsv, the same file taskActivation.py's
live analysis reads via conf/checkerboard_3cond.toml). The Psychtoolbox
equivalent is stimuli_ptb/checkerboard_3cond_task.m -- both present the exact
same events.tsv.

A full-contrast checkerboard flickers in one of three screen positions per
block: CENTER, LEFT, or RIGHT -- the same OFF (blank) -> ON (pattern A) ->
OFF -> ON (pattern B, the black<->white inverse of A) -> repeat flicker
checkerboard_1cond_task.py uses, at --flicker-hz. CENTER is a SQUARE (its height
set equal to its own width, not the window height) centered on screen, so
it stimulates only the fovea; LEFT and RIGHT are full window-height BARS,
anchored flush against the window's left/right edge respectively (not
floating with a gap), so they reach as far into the visual periphery as the
window allows. A small + fixation cross stays visible at screen center
THROUGHOUT every block (including LEFT/RIGHT, and every OFF phase), so the
subject can hold central gaze while the checkerboard stimulates each
position -- standard practice for a peripheral visual localizer, so the
resulting activation reflects retinotopic stimulus location rather than eye
movements.

Blocks: rest / center / rest / left / rest / right, x4 reps + a trailing
rest block, 12s blocks, 300s total -- matches conf/checkerboard_3cond.toml's
3-way one-vs-rest GLM contrast (glmCondA=center / glmCondB=left /
glmCondC=right, each contrasted against the mean of the other two) for
live analysis.

Shows a brief task-instructions screen (task description + the subject's
goal; dismiss with 1/2, or Escape to abort) before waiting for the scanner
trigger.

Run (needs `pip install psychopy` and a display):
    python checkerboard_3cond_task.py                    # waits for scanner trigger '5' or 't'
    python checkerboard_3cond_task.py --windowed          # not fullscreen, for testing
    python checkerboard_3cond_task.py --trigger-key space # press space yourself to start
-----------------------------------------------------------------------------"""
import os
import argparse
import datetime
import numpy as np
import common

HERE = os.path.dirname(os.path.realpath(__file__))
DEFAULT_EVENTS = os.path.join(common.PROJECT_ROOT, 'study_design', 'Checkerboard3Cond_events.tsv')


class _CenterLeftRightStim:
    """Draws the (possibly None) checker bar for this frame, THEN the
    fixation cross on top -- fixation must stay visible on EVERY frame of
    EVERY condition (see the module docstring), so a single stim_for()
    return value has to combine both rather than picking one or the other
    like the plain checkerboard_1cond_task.py / generic_motor_task.py do."""
    def __init__(self, checker, fixation):
        self.checker = checker
        self.fixation = fixation

    def draw(self):
        if self.checker is not None:
            self.checker.draw()
        self.fixation.draw()


def build_checker_stims(win, cells_across=8, center_width_frac=0.288, side_width_frac=0.32):
    """Build the two phase-inverted checkerboard GratingStims (pattern A /
    pattern B -- see the module docstring's OFF/A/OFF/B flicker) for each
    position (center/left/right), sized and placed as fractions of the
    window -- mirrors checkerboard_3cond_task.m's pixel-level destRects, but
    in PsychoPy 'height' units (window height = 1.0, width = aspect) so it
    scales to any window size the same way the rest of this project's
    PsychoPy stims do.

    CENTER is a SQUARE (height == center_width_frac, not the window height)
    centered on screen, so it stimulates only the fovea. LEFT and RIGHT
    span the FULL window height (vertical bars) and are only
    side_width_frac of the window wide; LEFT sits flush against the
    window's left edge and RIGHT flush against its right edge (no
    gap/margin) so each reaches as far into the periphery as the window
    allows. side_width_frac is still smaller than center_width_frac,
    matching the .m version.

    Returns {'center': {'A': stim, 'B': stim}, 'left': {...}, 'right': {...}}.

    THE CHECKER CELLS STAY SQUARE even though LEFT/RIGHT are tall, narrow
    rectangles. PsychoPy's GratingStim texture upload requires a SQUARE
    power-of-two array on some OpenGL backends (older/software renderers
    without the GL_ARB_texture_non_power_of_two extension) -- a non-square
    array like (n_rows, n_cols) with n_rows != n_cols logs a "Requiring a
    square power of two texture" error, and simply stretching a square
    array via size=(width, height) elongates the cells whenever
    width != height. So instead the texture stays a minimal 2x2 checker
    tile (still square, still power of two), and GratingStim's own
    spatial-frequency tiling (`sf`) repeats it `cells_across / 2` cycles
    horizontally and a DIFFERENT number of cycles vertically -- scaled by
    that stim's own height/width ratio -- so each repeated cell ends up
    the same physical height as width regardless of the stim's own aspect
    ratio (a no-op correction for CENTER, since it's already square)."""
    from psychopy import visual
    win_w_px, win_h_px = win.size
    aspect = win_w_px / win_h_px   # window width in 'height' units

    tile = np.array([[1.0, -1.0], [-1.0, 1.0]])   # minimal 2x2 checker tile
    tile_inv = -tile
    bar_height = 1.0   # full window height, for LEFT/RIGHT only
    sf_x = cells_across / 2.0   # cycles across the width (2 cells/cycle)

    def make(pos, width, height, arr):
        sf = (sf_x, sf_x * height / width)   # compensate for this stim's own aspect ratio
        return visual.GratingStim(win, tex=arr, mask=None, size=(width, height),
                                  pos=pos, units='height', interpolate=False, sf=sf)

    positions = {
        'center': (0.0, center_width_frac, center_width_frac),
        'left': (-aspect / 2 + side_width_frac / 2, side_width_frac, bar_height),
        'right': (aspect / 2 - side_width_frac / 2, side_width_frac, bar_height),
    }
    return {
        key: {'A': make((x, 0), width, height, tile), 'B': make((x, 0), width, height, tile_inv)}
        for key, (x, width, height) in positions.items()
    }


def main():
    ap = argparse.ArgumentParser(
        description='Present the CENTER/LEFT/RIGHT checkerboard localizer (PsychoPy).')
    ap.add_argument('--events', default=DEFAULT_EVENTS, help='events.tsv to present')
    ap.add_argument('--trigger-key', default='5,t',
                    help="comma-separated key(s) that start the run (scanner sync pulse, "
                         "or press one yourself to test)")
    ap.add_argument('--duration', type=float, default=None,
                    help='total run length in seconds (default: end of the last event); '
                         'pass nVols*TR to also show trailing rest')
    ap.add_argument('--flicker-hz', type=float, default=2.0,
                    help='how many times per second the checkerboard changes state during a '
                         'CENTER/LEFT/RIGHT block (default 2): OFF (blank) -> ON (pattern A) -> '
                         'OFF -> ON (pattern B, the black<->white inverse of A) -> repeat, each '
                         'state lasting 1/flicker_hz -- same semantics as checkerboard_1cond_task.py')
    ap.add_argument('--windowed', action='store_true', help='windowed instead of fullscreen')
    ap.add_argument('--log', default=None,
                    help='timing log path (default: stimuli/logs/checkerboard_3cond_<timestamp>.tsv)')
    args = ap.parse_args()

    from psychopy import visual, core

    events = common.read_events_tsv(args.events)
    print(f"[checkerboard_3cond] loaded {len(events)} events from {args.events}")

    log_path = args.log or os.path.join(
        HERE, 'logs', f"checkerboard_3cond_{datetime.datetime.now():%Y%m%d_%H%M%S}.tsv")

    win = visual.Window(fullscr=not args.windowed, color='black', units='height')
    checker_stims = build_checker_stims(win)
    # Green (not white) so it stays clearly visible against the
    # black-and-white checkerboard squares behind it.
    fixation = visual.TextStim(win, text='+', color='lime', height=0.1)
    flicker_hz = args.flicker_hz

    def stim_for(trial_type, t_in_event, duration):
        # True flicker: fully OFF (blank) between each ON flash, and the
        # flash itself alternates pattern A/B (black<->white inverse) so a
        # given screen location genuinely reverses polarity from one flash
        # to the next -- same 4-phase cycle as checkerboard_1cond_task.py.
        entry = checker_stims.get(trial_type)
        stim = None
        if entry is not None:
            phase = int(t_in_event * flicker_hz) % 4
            if phase == 1:
                stim = entry['A']
            elif phase == 3:
                stim = entry['B']
        return _CenterLeftRightStim(stim, fixation)

    instructions = (
        "CHECKERBOARD LOCALIZER TASK\n\n"
        "You will see a flickering black-and-white checkerboard pattern appear in "
        "different places on the screen -- sometimes in the middle, sometimes off "
        "to the left, sometimes off to the right.\n\n"
        "Goal: keep looking at the + in the center of the screen THE WHOLE TIME, "
        "even when the checkerboard appears off to one side -- just notice it out "
        "of the corner of your eye rather than looking directly at it. No response "
        "or button press is needed during the task.\n\n"
        "If you have any questions, ask the experimenter now. When you are "
        "comfortable, press any button to continue.")
    if common.show_instructions(win, instructions):
        core.quit()

    clock, t0 = common.wait_for_trigger(win, trigger_keys=args.trigger_key.split(','),
                                        instructions='Waiting for scanner trigger...')
    print(f"[checkerboard_3cond] triggered at t0={t0:.3f}s -- starting task")
    common.run_events(win, events, stim_for, clock, t0,
                      run_duration=args.duration, log_path=log_path)

    done = visual.TextStim(win, text='Task complete -- thank you!', color='white', height=0.06)
    done.draw()
    win.flip()
    core.wait(2.0)
    win.close()
    core.quit()


if __name__ == '__main__':
    main()
